### PR TITLE
ekf2: consolidate and simplify gnd effect logic

### PR DESCRIPTION
**Pull request by dagar**
I found the ekf2 gnd effect logic a bit convoluted and hard to follow with the somewhat awkward interaction where internal EKF state is updated when publishing estimator output. This is an attempt to make it a little simpler by pushing it entirely into the EKF backend.

Some of the logic is not identical, so might be worth a more careful review/discussion.

### CODE REVIEW

**Review by bresch — CHANGES_REQUESTED** [review]

File: src/modules/ekf2/EKF/ekf_helper.cpp [hunk 1]
```diff
@@ -1379,14 +1379,23 @@ void Ekf::updateBaroHgtBias()
...
-		if (isTimedOut(_time_last_gnd_effect_on, GNDEFFECT_TIMEOUT)) {
-
-			_control_status.flags.gnd_effect = false;
+void Ekf::updateGroundEffect()
+{
+	if (_control_status.flags.in_air && !_control_status.flags.fixed_wing) {
+		if (isTerrainEstimateValid()) {
+			// automatically set ground effect if terrain is valid
+			float height = _terrain_vpos - _state.pos(2);
+			_control_status.flags.gnd_effect = (height < _params.gnd_effect_max_hgt);
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/ekf2/EKF/ekf_helper.cpp:1388]: That was intentional, the idea being the timeout would only apply to the "manually" set gnd effect flag. So if terrain estimate became invalid you could still be in gnd effect if within the timeout (10s) of the external ground effect setting (takeoff, etc).

You think a 10 second timeout should still apply to terrain validity?

File: src/modules/ekf2/EKF/ekf.h [hunk 2]
```diff
@@ -301,6 +301,20 @@ class Ekf final : public EstimatorInterface
 	// use the latest IMU data at the current time horizon.
 	Quatf calculate_quaternion() const;
 
+	// set flag if static pressure rise due to ground effect is expected
+	// use _params.gnd_effect_deadzone to adjust for expected rise in static pressure
+	// flag will clear after GNDEFFECT_TIMEOUT uSec
+	void set_gnd_effect_flag(bool gnd_effect)
+	{
+		if (!isTerrainEstimateValid()) {
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/ekf2/EKF/ekf.h:309]: This was a last minute addition, the other idea was set_gnd_effect_flag() could simply enable it, then only the timeout would disable.

[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/ekf2/EKF/ekf.h]: How about the updated commit? https://github.com/PX4/PX4-Autopilot/pull/18868/commits/f482e39b2ba9f7a25cb2c2fb196ec805a9b598a2

 - the frontend can manually set gnd_effect based on external factors (like takeoff spool up) and it's disabled automatically after timeout
 - the backend terrain estimate always takes precedence

[the remark this answers is not in the document]
Comment (bresch) [reply src/modules/ekf2/EKF/ekf.h]: > How about the updated commit? f482e39

Yes, I find it much more understandable

File: src/modules/ekf2/EKF/ekf_helper.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bresch) [reply src/modules/ekf2/EKF/ekf_helper.cpp:1388]: Ok, no, I think this is fine

**Comment by dagar** [comment]
Conflicts with master resolved and rebased.

**Comment by dagar** [comment]
A lot of the relevant cleanup here happened along with moving the status flags to the delayed time horizon. https://github.com/PX4/PX4-Autopilot/pull/20576
